packages/apollo-server-core/src/utils/schemaManager.ts

```diff
@@ -0,0 +1,216 @@
+   * - If registered before start(), then the callback will be first called in
+   *   start(), and then later for updates.
+   * - If registered after start() but before stop(), the callback will be first
+   *   called in this method (for whatever the current schema is), and then
+   *   later for updates.
+   * - If registered after stop(), the callback will never be called.
+   *
+   * For gateways, a core supergraph SDL will be provided to the callback. If
+   * your gateway is too old to provide a core supergraph SDL, this method will
+   * throw.
+   *
+   * @param callback The listener to execute on schema load/updates.
+   */
+  public onSchemaLoadOrUpdate(
+    callback: (schemaContext: GraphQLSchemaContext) => void,
+  ): Unsubscriber {
+    if (
+      this.modeSpecificState.mode === 'gateway' &&
+      !this.modeSpecificState.gateway.onSchemaLoadOrUpdate
+    ) {
+      throw new Error(
+        'Your gateway is too old to register a onSchemaLoadOrUpdate() listener',
+      );
+    } else {
+      if (!this.isStopped && this.schemaContext) {
+        try {
+          callback(this.schemaContext);
+        } catch (e) {
+          this.logger.error(
+            [
+              "An error was thrown from an 'onSchemaLoadOrUpdate' listener:",
+              `${e?.message ?? e}`,
+            ].join(' '),
+          );
+        }
+      }
+      this.onSchemaLoadOrUpdateListeners.add(callback);
+    }
+
+    return () => {
+      this.onSchemaLoadOrUpdateListeners.delete(callback);
+    };
+  }
+
+  /**
+   * Get the schema-derived state for the current schema. This throws if called
+   * before start() is called.
+   */
+  public getSchemaDerivedData(): SchemaDerivedData {
+    if (!this.schemaDerivedData) {
+      throw new Error('You must call start() before getSchemaDerivedData()');
+    }
+    return this.schemaDerivedData;
+  }
+
+  /**
+   * Calling stop() will:
+   * - Stop gateway schema fetching (if a gateway was provided).
+   *   - Note that this specific step may not succeed if gateway is old.
+   * - Stop updating schema-derived data.
+   * - Stop notifying onSchemaLoadOrUpdate() listeners.
+   */
+  public async stop(): Promise<void> {
+    this.isStopped = true;
+    if (this.modeSpecificState.mode === 'gateway') {
+      this.modeSpecificState.unsubscribeFromGateway?.();
+      await this.modeSpecificState.gateway.stop?.();
+    }
+  }
+
+  private processSchemaLoadOrUpdateEvent(
+    schemaContext: GraphQLSchemaContext,
+  ): void {
+    if (!this.isStopped) {
+      this.schemaDerivedData = this.schemaDerivedDataProvider(
+        schemaContext.apiSchema,
+      );
+      this.schemaContext = schemaContext;
+      this.onSchemaLoadOrUpdateListeners.forEach((listener) => {
+        try {
+          listener(schemaContext);
+        } catch (e) {
+          this.logger.error(
+            [
+              "An error was thrown from an 'onSchemaLoadOrUpdate' listener:",
+              `${e?.message ?? e}`,
```

Comment:
Doesn't this lose the stacktrace?

Reply:
Good catch! I'm down to write it like
```typescript
this.logger.error(
  "An error was thrown from an 'onSchemaLoadOrUpdate' listener",
);
this.logger.error(e);
```
so we can keep the stacktrace (that seems to be the pattern at least for keeping stacktrace info, based on [this](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-core/src/ApolloServer.ts#L256-L257)).